File: tests/unit/core/test_reasoning.py
"""Tests for lamu.core.reasoning — think-block detection."""
from __future__ import annotations

import pytest

from lamu.core.reasoning import (
    NullReasoningExtractor,
    ReasoningExtractor,
    get_extractor,
)
from lamu.core.types import ReasoningMarker, StreamChunk


@pytest.fixture
def marker():
    return ReasoningMarker(open_tag="<think>", close_tag="</think>", family="qwen35")


@pytest.fixture
def extractor(marker):
    return ReasoningExtractor(marker)


def test_split_no_markers_returns_full_content(extractor):
    assert extractor.split("just a plain answer") == ("", "just a plain answer")


def test_split_extracts_reasoning(extractor):
    text = "<think>step by step</think>final answer"
    reasoning, content = extractor.split(text)
    assert reasoning == "step by step"
    assert content == "final answer"


def test_split_truncated_no_close(extractor):
    text = "<think>still thinking, hit token limit"
    reasoning, content = extractor.split(text)
    assert reasoning == "still thinking, hit token limit"
    assert content == ""


def test_strip_returns_only_content(extractor):
    assert extractor.strip("<think>x</think>y") == "y"


def test_stream_filter_no_reasoning_block(extractor):
    chunks = list(extractor.stream_filter(iter(["hello ", "world"])))
    text = "".join(c.text for c in chunks)
    assert "hello" in text
    assert all(c.type == "content" for c in chunks)


def test_stream_filter_drops_reasoning_by_default(extractor):
    tokens = ["<think>", "stuff ", "more", "</think>", "answer"]
    chunks = list(extractor.stream_filter(iter(tokens), include_reasoning=False))
    content = "".join(c.text for c in chunks if c.type == "content")
    assert "stuff" not in content
    assert "answer" in content


def test_stream_filter_includes_reasoning_when_requested(extractor):
    tokens = ["<think>", "stuff", "</think>", "answer"]
    chunks = list(extractor.stream_filter(iter(tokens), include_reasoning=True))
    types = [c.type for c in chunks]
    assert "reasoning" in types
    assert "content" in types


def test_null_extractor_passthrough():
    e = NullReasoningExtractor()
    assert e.split("hi") == ("", "hi")
    assert e.strip("hi") == "hi"
    chunks = list(e.stream_filter(iter(["a", "b"])))
    assert [c.text for c in chunks] == ["a", "b"]
    assert all(c.type == "content" for c in chunks)


def test_get_extractor_picks_correct_type(marker):
    assert isinstance(get_extractor(marker), ReasoningExtractor)
    assert isinstance(get_extractor(None), NullReasoningExtractor)


def test_stream_filter_buffer_cap_raises(extractor):
    """Phase C: unbounded think buffer raises ReasoningOverflow at 64 KB."""
    from lamu.core.errors import ReasoningOverflow
    huge_token = "x" * 100_000
    with pytest.raises(ReasoningOverflow):
        list(extractor.stream_filter(iter(["<think>", huge_token])))

File: lamu-rs/lamu-core/tests/test_reasoning.rs
use lamu_core::reasoning::{get_extractor, MarkerExtractor, NullExtractor, ReasoningExtractor};
use lamu_core::types::{ReasoningMarker, StreamChunk};

fn marker() -> ReasoningMarker {
    ReasoningMarker {
        open_tag: "<think>".to_string(),
        close_tag: "</think>".to_string(),
        family: "qwen35".to_string(),
    }
}

#[test]
fn split_basic() {
    let ext = MarkerExtractor::new(marker());
    let (r, c) = ext.split("<think>\nhmm let me think\n</think>\nThe answer is 42.");
    assert_eq!(r, "hmm let me think");
    assert_eq!(c, "The answer is 42.");
}

#[test]
fn split_no_marker() {
    let ext = MarkerExtractor::new(marker());
    let (r, c) = ext.split("just plain text");
    assert_eq!(r, "");
    assert_eq!(c, "just plain text");
}

#[test]
fn split_truncated() {
    let ext = MarkerExtractor::new(marker());
    let (r, c) = ext.split("<think>\nstill thinking...");
    assert_eq!(r, "still thinking...");
    assert_eq!(c, "");
}

#[test]
fn null_passthrough() {
    let ext = NullExtractor;
    let (r, c) = ext.split("anything");
    assert_eq!(r, "");
    assert_eq!(c, "anything");
}

#[test]
fn factory() {
    let ext = get_extractor(Some(marker()));
    let (r, _) = ext.split("<think>x</think>y");
    assert_eq!(r, "x");

    let ext = get_extractor(None);
    let (r, c) = ext.split("anything");
    assert_eq!(r, "");
    assert_eq!(c, "anything");
}

#[test]
fn stream_strip() {
    let ext = MarkerExtractor::new(marker());
    let tokens: Vec<String> = vec!["<think>".into(), "thinking ".into(), "tokens".into(),
        "</think>".into(), "Hi".into(), " there".into()];
    let chunks: Vec<StreamChunk> = ext.stream_filter(Box::new(tokens.into_iter()), false).collect();
    let content: String = chunks.iter().filter_map(|c| match c {
        StreamChunk::Content(s) => Some(s.as_str()),
        _ => None,
    }).collect();
    assert!(content.contains("Hi"));
    assert!(!content.contains("thinking"));
}

#[test]
fn stream_include_reasoning() {
    let ext = MarkerExtractor::new(marker());
    let tokens: Vec<String> = vec!["<think>".into(), "abc".into(), "</think>".into(), "ok".into()];
    let chunks: Vec<StreamChunk> = ext.stream_filter(Box::new(tokens.into_iter()), true).collect();
    let has_reasoning = chunks.iter().any(|c| matches!(c, StreamChunk::Reasoning(_)));
    assert!(has_reasoning, "expected reasoning chunk");
}
